### ISSUE TITLE
☂️ Analyzer / Linter Infrastructure

### ISSUE THREAD

**Issue by leops** [rome/tools#2561]
# Description

Deliver a baseline version of the analyzer implementing a few basic rules, infrastructure for emitting code actions / auto-fixes, and language server as well as CLI integration

# Analyzer

- [x] Syntax Factory / Mutation API + `analyze` crate refactor: #2491
- [x] Establish the (probably large) list of rules we need to implement to consider the linter ready for beta

# CLI

- [x] Generalize the use of markup to all diagnostics: #2580
- [x] CLI integration (check command): #2583
- [x] `--no-colors` / `atty` check to disable printing markup: #2613
- [x] `--fix` argument  to run all safe code fixes: #2615
- [ ] ~`--review` argument to pass to `rome check` command, to manually accept proposed fixes~

# Language Server

- [x] Improved editor integration: #2576
- [x] Investigate an "auto-fix all" command in the editor: #2615

**Comment by leops** [comment]
Currently all the tasks for the first analysis milestone are done, I think I kept this open in case we wanted to add tasks from a future milestone in there too if they were related to the "delivering an initial version of linting" umbrella, but we haven't decided on that yet

**Comment by ematipico** [comment]
So we have now created a huge list of rules that we could implement, although the list is big and some of them require a discussion.

This makes it difficult to understand or cut a first version of the linter.

I'd propose the following: the first version of the linter should only contain rules coming from old Rome, minus the ones that require scoping information. The old rules were reviewed in terms of mission and documentation.

The `--review` argument is nice to have, but I think we can live without it for the initial release.

This would gives us a finish line before starting pushing new rules. (Some coming from clippy, others from eslint, etc.).

In light of this, I would suggest to break down the huge list in different umbrellas: one with the rules coming from Rome minus the ones that need scope information (these will be rules to ship first), and  an umbrella with the Rome rules that need scoping (so we don't lose them), and one umbrella with the propose rules from different sources (eslint, clippy, etc.). What do you think?

To me, it seems a good plan and it won't make external contributions loose. I believe it's a good plan.

**Comment by ematipico** [comment]
There's also the matter of diagnostics. We are creating them inside the linter, but I believe the end plan is to move ALL diagnostics in one single place, like we did with Rome classic.

I'm not sure how to tackle it, but I think we should it sooner rather than later, because lint rules will generate LOT OF diagnostics. The refactor might take more time, the more diagnostics we have to move.


In light of that, could share what are the next steps we have to make in order to achieve the end plan?

**Comment by ematipico** [comment]
Thank you Leo! 

I will make sure the break down those umbrellas, so we can have clear goals.

I leave the status of this issue to you. It seems we implemented all the required tasks of this umbrella, and it seems that diagnostics are a bigger picture so they are not related to linter only.
